### DISCUSSION TITLE
New Demo: BPMN Pools, Swimlanes & Milestones

### DISCUSSION THREAD

**Discussion by kumilingus** [clientIO/joint#2396] [News]
We're working towards no-dependency _JointJS_. But that's not all. We're building other features as well, and we thought it would be a good idea to give you a taste now.

The existing implementation of **BPMN pools** has fallen short of meeting the actual requirements of a BPMN application. Specifically, the **elements** lacked **semantic association with lanes**.

Another problem we tried to address is the necessity to **configure the swimlanes and milestones via** a form e.g. Inspector. We realize that users prefer a more user-friendly **drag-and-drop  approach**.

Here it is. The features of new version of JointJS BPMN Pools.

1. **Working with swimlanes**
   - inserting new swimlanes in drag-and-drop fashion
   - swimlanes reordering in drag-and-drop fashion
   - automatic swimlane resizing to accommodate the content.
   - swimlane resizing in any direction
    <img width="600" src="https://github.com/clientIO/joint/assets/3967880/99f762b6-fdde-4810-9346-badacfdbb150" />

2. **Working with phases (milestones)**
    - insert of phases in drag-and-drop fashion
    - automatic phases resizing to accommodate the content.
    - phase resizing in any direction 
    <img width="600" src="https://github.com/clientIO/joint/assets/3967880/b9832d54-8d58-4256-867b-0a7b83caf1f0" />

3.  **Content awereness**
    - prevent the user from resizing swimlanes & milestones beyond the content boundary
    <img width="600" src="https://github.com/clientIO/joint/assets/3967880/4850adf8-27c5-4681-b94d-bac53ff09c22" />

4. **Make it clear what will be the result of any drag & drop action**
    - always showing previews
    - highlighting target lanes and phases for the drop
    <img width="600" src="https://github.com/clientIO/joint/assets/3967880/347349ed-911e-48c7-9ceb-169c825e3d95" />

5.  **Moving lanes between different pools**
    <img width="600" src="https://github.com/clientIO/joint/assets/3967880/4a899964-4c45-4966-84f5-3adab15ef38d" />

6.  **Inline editing of all text**
    <img width="600" src="https://github.com/clientIO/joint/assets/3967880/6dda8a7a-58ad-4e45-9084-73d250475f85" />

And here is the [live BPMN Pools demo](https://resources.jointjs.com/demos/rappid/apps/BPMNPools/) where you can experience all above for yourself.

_Note: We're planning to introduce also vertical version of the same._

**Comment by shivamfxbytes** [comment]
Hii kumilingus, but how to get the source code and json object for the render BPMN pool

**Reply by kumilingus** [reply]
If you are an existing customer with valid license you can send us an email to support@client.io and we can share the code with you.
